File: go/internal/sol/p0060/p0060.go
package p0060

import (
	"slices"
	"strconv"

	"pe-solver/internal/mylib"
)

type intSet map[int]struct{}

func pairablePrimes(x int, ascPrimes []int, limit int) []int {
	upper_x := mylib.Pow(10, mylib.NumOfDigits(x, 10))
	upper_p := 10
	result := make([]int, 0)
	for _, p := range ascPrimes {
		if x+p >= limit {
			break
		}
		for p > upper_p {
			upper_p *= 10
		}
		if mylib.IsPrime(x*upper_p+p) && mylib.IsPrime(p*upper_x+x) {
			result = append(result, p)
		}
	}

	return result
}

func findCliques(ascPrimes []int, size int, tbl map[int]intSet) [][]int {
	isAllPair := func(p int, group []int) bool {
		for _, x := range group {
			if _, ok := tbl[x][p]; !ok {
				return false
			}
		}

		return true
	}

	descPrimes := slices.Clone(ascPrimes)
	slices.Reverse(descPrimes)
	result := make([][]int, 0)

	var aux func([]int, []int, int)
	aux = func(group, descPrimes []int, depth int) {
		if depth == 0 {
			result = append(result, group)
		} else {
			for idx, p := range descPrimes[:len(descPrimes)-depth+1] {
				if len(group) == 0 || isAllPair(p, group) {
					aux(append(group, p), descPrimes[idx+1:], depth-1)
				}
			}
		}
	}

	aux(make([]int, 0), descPrimes, size)

	return result
}

func compute(groupSize int) string {
	primeGroups := [][]int{{3}, {3}}
	tbl := map[int]intSet{}
	size := groupSize - 1
	result := int(^uint(0) >> 1) // maxInt

	// skip 2, 3 and 5, and start with 7
	for p := 7; p < result; p = mylib.NextPrime(p) {
		grpIdx := (p % 3) - 1
		nbrs := pairablePrimes(p, primeGroups[grpIdx], result)
		numSet := intSet{}
		for _, x := range nbrs {
			numSet[x] = struct{}{}
		}
		tbl[p] = numSet
		primeGroups[grpIdx] = append(primeGroups[grpIdx], p)

		if len(nbrs) < size {
			continue
		}

		cliques := findCliques(nbrs, size, tbl)
		if len(cliques) > 0 {
			for _, clq := range cliques {
				acc := p
				for _, x := range clq {
					acc += x
				}
				result = min(result, acc)
			}
		}
	}

	return strconv.FormatInt(int64(result), 10)
}

func Solve() string {
	return compute(5)
}

File: haskell/src/Sol/P0060.hs
module Sol.P0060 (compute, solve) where

import Control.Applicative (Alternative, empty, (<|>))
import Data.List (tails, uncons)
import Data.Maybe (mapMaybe)

import qualified Data.IntMap as M (IntMap, empty, insert, (!))
import qualified Data.IntSet as S (IntSet, fromList, member)

import Mylib.Math (numOfDigits)
import Mylib.Prime (isPrime, primeNumbers)

-- [3, 7, 13, 19, ...]
primes_rem1 :: [Int]
primes_rem1 = 3 : filter ((== 1) . (`mod` 3)) (drop 3 primeNumbers)

-- [3, 11, 17, 23, ...]
primes_rem2 :: [Int]
primes_rem2 = 3 : filter ((== 2) . (`mod` 3)) (drop 3 primeNumbers)

pairablePrimes :: Int -> Int -> [Int]
pairablePrimes p limit =
    [x | x <- takeWhile (\n -> n + p < limit && n < p) ascPrimes, isPair x]
  where
    upper_p = 10 ^ numOfDigits p 10
    ascPrimes =
        if p `mod` 3 == 1
            then primes_rem1
            else primes_rem2

    isPair :: Int -> Bool
    isPair x =
        isPrime (x * upper_p + p) && isPrime (p * upper_x + x)
      where
        upper_x = 10 ^ numOfDigits x 10

findCliques :: [Int] -> Int -> M.IntMap S.IntSet -> [[Int]]
findCliques dscNbrs targetSize tbl =
    dfs findClqs [([], dscNbrs)]
  where
    dfs ::
        Alternative m =>
        (([Int], [Int]) -> [([Int], [Int])]) -> [([Int], [Int])] -> m [Int]
    dfs _ [] = empty
    dfs f (x@(clq, _) : xs)
        | length clq == targetSize =
            pure clq <|> dfs f xs
        | otherwise =
            dfs f (f x ++ xs)

    findClqs :: ([Int], [Int]) -> [([Int], [Int])]
    findClqs (clq, nbrs) =
        map (\(n, lst) -> (n : clq, lst))
            . filter (\(n, _) -> all (S.member n . (tbl M.!)) clq)
            $ cands
      where
        cands =
            mapMaybe
                (\lst -> if length lst >= targetSize - length clq then uncons lst else Nothing)
                $ tails nbrs

compute :: Int -> String
compute groupSize =
    show $ aux maxBound M.empty (drop 3 primeNumbers)
  where
    aux :: Int -> M.IntMap S.IntSet -> [Int] -> Int
    aux _ _ [] = error "fatal error (unreachable)"
    aux minSum tbl (p : ps)
        | p >= minSum =
            -- All prime numbers under the `minSum` have been checked,
            -- and we've confirmed that it is the smallest sum.
            minSum
        | otherwise =
            -- Otherwise, the `minSum` is not confirmed that it's the
            -- smallest sum yet. So, continue searching for cliques.
            let nbrs = pairablePrimes p minSum
                new_tbl = M.insert p (S.fromList nbrs) tbl
             in if length nbrs < groupSize - 1
                    then
                        aux minSum new_tbl ps
                    else case findCliques (reverse nbrs) (groupSize - 1) tbl of
                        [] -> aux minSum new_tbl ps
                        cliques -> aux (min minSum (getMinSumOfClqs p cliques)) new_tbl ps
      where
        getMinSumOfClqs :: Int -> [[Int]] -> Int
        getMinSumOfClqs x clqs = x + minimum (sum <$> clqs)

solve :: String
solve = compute 5
